db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/SelectNode.java

```diff
@@ -1601,39 +1599,37 @@ public ResultSetNode optimize(DataDictionary dataDictionary,
                 }
             }
         }
-		/* Get a new optimizer */
-        optimizer=getOptimizer(fromList, wherePredicates, dataDictionary, orderByList);
-        optimizer.setOuterRows(outerRows);
 
-        // Aggregation with no GROUP BY always outputs one row.
-        if ((selectAggregates != null && !selectAggregates.isEmpty() ||
-             havingAggregates !=null  && !havingAggregates.isEmpty()) && hasNoGroupBy())
-            optimizer.setSingleRow(true);
+        Optimizer optimizer = getOptimizer(fromList, wherePredicates, dataDictionary, orderByList);
+        if (nestingLevel == 0) {
+            // Top level select node, we should try to cost control first, and fallback to spark if necessary
 
-		/* Optimize this SelectNode */
-        while(optimizer.nextJoinOrder()){
-            while(optimizer.getNextDecoratedPermutation()){
-                optimizer.costPermutation();
+            if (!determineSpark()) {
+                optimizer.setForSpark(false);
+                findBestPlan(optimizer, dataDictionary, outerRows);
+            }
+
+            if (determineSpark()) {
+                for (int i = 0; i < fromList.size(); ++i) {
+                    FromTable ft = (FromTable) fromList.elementAt(i);
+                    ft.resetAccessPaths();
```

Comment:
This seems only resetting the access path for the tables in the top level SELECT, but not that of the nested SELECT, shouldn't we reset that too?

Reply:
Done